### PR TITLE
chore: Record the duration metric for the authz call.

### PR DESCRIPTION
**Pull request by wiedld**
create DurationHistogramDecorator, and utilize to record the authz duration in SingleTenantRequestUnifier

---
* chore: create DurationHistogramDecorator in metric mod (1904035e12c86851de9e1a0e6946d5abf65e86fa)
* chore: impl Authorizer for DurationMetricDecorator (51c4597ef2da03bf31f704ceb7aa13a3b92e3303)
* chore: wrap authz provided to SingleTenantRequestUnifier with DurationMetricDecorator (cadc07fb8e6dcfdfa48475b053b9eb4a81d1d24c)

## Checklist
- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: metric/src/decorators.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            inner,
+            time_provider: Default::default(),
+            success,
+            error,
+        }
+    }
+
+    pub fn inner(&self) -> &T {
+        &self.inner
+    }
```
Comment (domodwyer) [inline metric/src/decorators.rs:56]: So normally a decorator doesn't _look like_ a decorator - there's no concept of an `inner` or that the code is wrapped in any way.

I think if you hoist the `record_duration()` body up to the `Authoriser` impl, then this, and the `record_duration()` can be removed too - your authoriser impl will be able to call `self.inner` directly :+1:

File: metric/src/decorators.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    error: DurationHistogram,
+}
+
+impl<T> DurationHistogramDecorator<T> {
+    /// Wrap a new [`DurationHistogramDecorator`] over `T` exposing metrics.
+    pub fn new(
+        title: &'static str,
+        description: &'static str,
+        registry: &Registry,
+        inner: T,
```
[the remark this answers is not in the document]
Comment (domodwyer) [reply metric/src/decorators.rs:39]: The default is fine 👌 

I think we should push these the title/description down into the constructor body too!

**Review by domodwyer — CHANGES_REQUESTED** [review]
I see what you were going for!

The downside of this one-decortor-to-rule-them-all approach is that it has _one metric_ inside - if authoriser added another method we wanted to instrument, this generalised decorator can only address a single metric.

Imagine decorating an interface like this:

```rust
trait Thinger {
    async fn A(); // metric name: thinger_a_latency_histogram
    async fn B(); // metric name: thinger_b_latency_histogram
}
```

Then you'd need this decorator to have two metrics internally. Or then three, four, etc.

While it was definitely a good exercise, I think the [specialised wrapper per interface][example] is the way to go :+1:

[example]:https://github.com/influxdata/influxdb_iox/blob/34bb7d8100b14e8bf03eacc2d7a41c7707e7dd84/ingester2/src/dml_sink/instrumentation.rs#L45-L68

File: metric/src/decorators.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wiedld) [reply metric/src/decorators.rs:39]: done.

**Comment by wiedld** [comment]
> While it was definitely a good exercise, I think the [specialised wrapper per interface](https://github.com/influxdata/influxdb_iox/blob/34bb7d8100b14e8bf03eacc2d7a41c7707e7dd84/ingester2/src/dml_sink/instrumentation.rs#L45-L68) is the way to go 👍

Closing this PR, in favor of [another PR](https://github.com/influxdata/influxdb_iox/pull/7690) which has the approach^^ suggested.

**Pull request closed without merging**